File: M3104/tomcat-etud/webapps/gpsapp/WEB-INF/src/db/CoordinatesDAO.java
package db;


import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class CoordinatesDAO implements DAO<Coordinates> {
    private static CoordinatesDAO inst;
    private db.SqliteConnection connection = db.SqliteConnection.getInstance();

    public static CoordinatesDAO getInstance() {
        if (CoordinatesDAO.inst == null) {
            CoordinatesDAO.inst = new CoordinatesDAO();
        }
        return CoordinatesDAO.inst;
    }

    public int insert(Coordinates coordinates) {
        int newId = -1;
        Connection con = null;

        String query = "INSERT INTO Coordinates"
                + "(journey_id, journey_pos, latitude, longitude) VALUES"
                + "(?,?,?,?)"
                + ";";

        String counter = "SELECT COUNT(*) FROM Coordinates;";

        try {
            con = connection.getConnection();
            int rows = con.prepareStatement(counter).executeQuery().getInt(1);
            PreparedStatement state = con.prepareStatement(query);

            state.setInt(1, rows+1);
            state.setInt(2, coordinates.getJourney_pos());
            state.setDouble(3, coordinates.getLatitude());
            state.setDouble(4, coordinates.getLongitude());

            state.executeUpdate();
            newId = rows + 1;
            coordinates.setJourney_id(newId);

        } catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }

        return  newId;
    }

    public boolean update(Coordinates coordinates) {
        boolean ret = false;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "UPDATE Coordinates SET journey_pos = ?"
                    + ", latitude = ?"
                    + ", longitude = ?"
                    + " WHERE journey_id = ?";
            PreparedStatement state = con.prepareStatement(query);

            state.setInt(1, coordinates.getJourney_pos());
            state.setDouble(2, coordinates.getLatitude());
            state.setDouble(3, coordinates.getLongitude());
            state.setInt(4, coordinates.getJourney_id());

            state.executeUpdate();
            ret = true;
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    public boolean delete(Coordinates coordinates) {
        boolean ret = false;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "DELETE FROM Coordinates "
                    + " WHERE journey_pos = ?"
                    + ";";
            PreparedStatement state = con.prepareStatement(query);
            state.setInt(1,  coordinates.getJourney_pos());
            state.executeUpdate();
            ret = true;
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    public Coordinates find(int id) {
        Coordinates coordinates = null;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "SELECT * FROM Coordinates WHERE journey_id = ?";
            PreparedStatement state = con.prepareStatement(query);
            state.setInt(1, id);
            ResultSet res = state.executeQuery();
            coordinates = new Coordinates(
                    res.getInt("journey_pos"),
                    res.getDouble("latitude"),
                    res.getDouble("longitude")
            );
            coordinates.setJourney_id(res.getInt("journey_id"));
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return coordinates;
    }


}

File: M3104/PHP/my_gps_web_app/model/CoordinatesDAO.php
<?php
require_once('SqliteConnection.php');
require_once("Coordinates.php");
require_once("JourneyDAO.php");
require_once("Journey.php");
require_once("controllers/CalculDistanceImpl.php");

class CoordinatesDAO {
    private static $dao;

    private function __construct() {
    }

    public final static function getInstance() {
        if(!isset(self::$dao)) {
            self::$dao= new CoordinatesDAO();
        }
        return self::$dao;
    }

    public final function find($id) {
        $dbc = SqliteConnection::getInstance()->getConnection();
        $stmt = $dbc->prepare("SELECT * FROM Coordinates WHERE journey_pos = :journey_pos");
        $stmt->bindValue(':journey_pos', (int) $id, PDO::PARAM_INT);
        $stmt->execute();
        $result = $stmt->fetch(PDO::FETCH_ASSOC);

        $coordinates = new Coordinates();
        $coordinates->setJourneyId($result["journey_id"]);
        $coordinates->setJourneyPos($result["journey_pos"]);
        $coordinates->setLatitude($result["latitude"]);
        $coordinates->setLongitude($result["longitude"]);
        return $coordinates;
    }

    public final function findWithJourneyId($id) {
        $dbc = SqliteConnection::getInstance()->getConnection();
        $stmt = $dbc->prepare("SELECT * FROM Coordinates WHERE journey_id = :id");
        $stmt->bindValue(':id', (int) $id, PDO::PARAM_INT);
        $stmt->execute();
        $results = $stmt->fetchALL(PDO::FETCH_CLASS, 'Coordinates');
        return $results;
    }

    public final function findAll(){
        $dbc = SqliteConnection::getInstance()->getConnection();
        $query = "SELECT * FROM Coordinates";
        $stmt = $dbc->query($query);
        $results = $stmt->fetchALL(PDO::FETCH_CLASS, 'Coordinates');
        return $results;
    }

    public final function insert(Coordinates $coordinates){
        if($coordinates instanceof Coordinates){
            $dbc = SqliteConnection::getInstance()->getConnection();
            $dbc->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
            $stmt = $dbc->prepare("INSERT INTO Coordinates(journey_id, journey_pos, latitude, longitude) VALUES (:journey_id, :journey_pos, :latitude, :longitude)");
            $stmt->bindValue(':journey_id', $coordinates->getJourneyId(), PDO::PARAM_STR);
            $stmt->bindValue(':journey_pos', $coordinates->getJourneyPos(), PDO::PARAM_STR);
            $stmt->bindValue(':latitude', $coordinates->getLatitude(), PDO::PARAM_STR);
            $stmt->bindValue(':longitude', $coordinates->getLongitude(), PDO::PARAM_STR);
            $stmt->execute();
            $this->updateDistance($coordinates->getJourneyId());
            return true;
        }
        else {
            return false;
        }
    }

    public final function getLastId() {
        $dbc = SqliteConnection::getInstance()->getConnection();
        $query = "SELECT Count(*) FROM Coordinates";
        $stmt = $dbc->query($query);
        $result = $stmt->fetch();
        return $result[0];
    }

    public final function delete(Coordinates $coordinates) {
        if($coordinates instanceof Coordinates) {
            $dbc = SqliteConnection::getInstance()->getConnection();
            $stmt = $dbc->prepare("DELETE FROM Coordinates WHERE journey_pos = :journey_pos");
            $stmt->bindValue(':journey_pos', $coordinates->getJourneyPos(), PDO::PARAM_INT);
            $stmt->execute();
            $c = $this->find($coordinates->getJourneyPos());
            $this->updateDistance($c->getJourneyId());
            return true;
        }
        else {
            return false;
        }
    }

    public final function update(Coordinates $coordinates) {
        if($coordinates instanceof Coordinates) {
            $dbc = SqliteConnection::getInstance()->getConnection();
            $stmt = $dbc->prepare("UPDATE Coordinates SET journey_pos = :journey_pos, latitude = :latitude, longitude = :longitude WHERE journey_id = :journey_id");
            $stmt->bindValue(':journey_pos', $coordinates->getJourneyPos(), PDO::PARAM_INT);
            $stmt->bindValue(':latitude', $coordinates->getLatitude(), PDO::PARAM_STR);
            $stmt->bindValue(':longitude', $coordinates->getLongitude(), PDO::PARAM_STR);
            $stmt->bindValue(':journey_id', $coordinates->getJourneyId(), PDO::PARAM_INT);
            $stmt->execute();
            $this->updateDistance($coordinates->getJourneyId());
            return true;
        }
        else {
            return false;
        }
    }

    public final function updateDistance($id) {
        $id = (int) $id;
        $coords = $this->findWithJourneyId($id);
        $latlong = array();
        foreach ($coords as $c) {
            $latlong[] = array($c->getLatitude(), $c->getLongitude());
        }

        $journeyDao = JourneyDAO::getInstance();
        $journey = $journeyDao->find($id);

        $calcul = new CalculDistanceImpl();
        $distance = $calcul->calculDistanceTrajet($latlong);
        $journey->setDistance($distance);
        $journeyDao->update($journey);
    }
}
?>
